versions/3.0.4.md

```diff
@@ -3523,3 +3527,215 @@ Version   | Date       | Notes
+
+Here is one such template.
+The exact variable names are unimportant, but in this example "f" and "w" are meant to suggest "formulas" and "words", while "n" and "v" suggest "name" and "value":
+
+```urlencoded
+?{+fn0}={+fv0}&{+fn1}={+fv1}&{+fn2}={+fv2}&{wn}={wv0} {wv1} {wv2}
+```
+
+We'll also need to pre-process the names and values for `formaulas` because while `/` and most other reserved characters are allowed in the query string by RFC3986, `[`, `]`, and `#` [are not](https://datatracker.ietf.org/doc/html/rfc3986#appendix-A), and `&`, `=`, and `+` all have [special behavior](https://www.rfc-editor.org/rfc/rfc1866#section-8.2.1) in the `application/x-www-form-urlencoded` format, which is what we are using in the query string.
+
+Setting `allowReserved: true` does _not_ make reserved characters that are not allowed in URIs allowed, it just allows them to be _passed through expansion unchanged._
+Therefore, any tooling still needs to percent-encode those characters because reserved expansion will not do it, but it _will_ leave the percent-encoded triples unchanged.
+See also Appendix D for further guidance on percent-encoding and form media types, including guidance on handling the delimiter characters for `spaceDelimited`, `pipeDelimited`, and `deepObject` in parameter names and values.
+
+So here is our data structure that arranges the names and values to suit the template above, where names and values for `formulas` have `[]#&=+` pre-percent encoded (although only `+` appears in this example):
+
+```YAML
+fn0: a
+fv0: x%2By
+fn1: b
+fv1: x/y
+fn2: c
+fv2: x^y
+wn: words
+wv0: math
+wv1: is
+wv2: fun
+```
+
+Expanding our manually assembled template with our restructured data yields the following query string:
+
+```urlencoded
+?a=x%2By&b=x/y&c=x%5Ey&words=math%20is%20fun
+
+```
+The `/` and the pre-percent-encoded `%2B` have been left alone, but the disallowed `^` character (inside a value) and space characters (in the template but outside of the expanded variables) were percent-encoded.
+
+#### Undefined Values and Manual URI Template Construction
+
+Care must be taken when manually constructing templates to handle the values that [RFC6570 considers to be _undefined_](https://datatracker.ietf.org/doc/html/rfc6570#section-2.3) correctly:
+
+```YAML
+formulas: {}
+words:
+- hello
+- world
+```
+
+Using this data with our original RFC6570-friendly URI Template, `{?formulas*,words}`, produces the following:
+
+
+```urlencoded
+?words=hello,world
+```
+
+This means that the manually constructed URI Template and restructured data need to leave out the `formulas` object entirely so that the `words` parameter is the first and only parameter in the query string.
+
+Restructured data:
+
+```YAML
+wn: words
+wv0: hello
+wv1: world
+```
+
+Manually constructed URI Template:
+
+```urlencoded
+?{wn}={wv0} {wv1}
```

Comment:
I'm not sure I understand the advantage of that manually constructed URI template over

> ?words={wv0} {wv1}

Reply:
It's a good question and relates to your other two comment about [not needing to set `allowReserved` for names with percent encoding](https://github.com/OAI/OpenAPI-Specification/pull/3818#discussion_r1616048516) and [not needing the `+` on the formula names](https://github.com/OAI/OpenAPI-Specification/pull/3818#pullrequestreview-2080933650).

My original thought was "simulate the normal behavior of parameter names in `form` by putting them through the same process as the values" which led to "if you need pre-percent encoded values in names then you'll have to set `allowReserved: true` to get them.

But since that seems not to be true (two Python URI Template implementations both agree with you on that), then there's no need to use `allowReserved` and therefore no need to make the names template variables since they have to be pre-encoded anyway.  It's simpler to just copy them as part of the literal portion of the template.  I'll make an update taking all of this into account.